### PR TITLE
feat: Replace JClouds with AWS SDK + File NIO [DHIS2-20648]

### PR DESCRIPTION
**Pull request by david-mackessy**
AI assisted 🤖

# Summary

Replaces the JClouds object storage implementation. This was primarily planned due to the [archiving of the JCLouds library](https://attic.apache.org/projects/jclouds.html). Replacing it has resulted in many positives:
- introduction of `BlobStoreService` interface
- dedicated handling of different `FileStoreProvider` types
- simpler extension path for different `FileStoreProvider`

JClouds tried to handle all variants of `FileStoreProvider` (noted in the table below) but it did so in an inefficient way:
- needed to know type of `FileStoreProvider` to decide what to do
- separate code paths for different `FileStoreProvider` types
- hard to make changes for one `FileStoreProvider` type without affecting others

New implementations
| Provider | New Implementation |
|--|--|
| aws-s3 | AWS SDK v2 |
| s3 | AWS SDK v2 |
| filesystem | Java File NIO |
| transient | Java ConcurrentHashMap |

## Rename
- Replaced any class with `JClouds` prefix with `BlobStore`
- `AppStorageSource.JCLOUDS` ->  `AppStorageSource.BLOB_STORE`

## Removed
`JCloudsStore` implementation

## Added
Implementations for:
- `S3BlobStoreService`
- `FileSystemBlobStoreService`
- `TransientBlobStoreService`

## Fix
As part of the replacement, recursive deletes now operate correctly. Previous issue of some files being left behind depending on provider type. Haven't seen any evidence of this while testing.

# Testing

## Automated
- Unit tests
- Integration tests

## Manual
Tested providers (`s3`, `filesystem`, `transient`) with the scenarios:

| Scenario | Result |
| -- | -- |
| start server with no bundled apps installed | all apps installed |
| browse through apps in UI | no issues loading |
| start server with no default icons | `PATCH /api/icons` -> icons installed |
| retrieve icon `2g_negative` | icon retrieved |
| upload data value file | file saved |
| import metadata async | job data saved |
| update existing bundled app | deleted old version & installed new version |
| upload custom app | installed |
| get `index.html` from custom app | retrieved |
| delete custom app | deleted |

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java [hunk 1]
```diff
@@ -33,5 +33,5 @@
  * @author Stian Sandvold
  */
 public enum AppStorageSource {
-  JCLOUDS
+  BLOB_STORE
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java]: We need to check if values of this enum are stored somewhere. I am not sure.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java [hunk 2]
```diff
@@ -79,8 +79,8 @@
  */
 @Slf4j
 @RequiredArgsConstructor
-@Service("org.hisp.dhis.appmanager.JCloudsAppStorageService")
-public class JCloudsAppStorageService implements AppStorageService {
+@Service("org.hisp.dhis.appmanager.BlobStoreAppStorageService")
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java]: If the name can change we maybe can also remove the explicit naming?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java [hunk 3]
```diff
@@ -338,30 +344,17 @@ public void deleteApp(@Nonnull App app) {
...
+    blobStore.deleteDirectory(folder.asPrefix());
     log.info("Deleted app {}", app.getName());
   }
 
   @Override
   @Nonnull
   public ResourceResult getAppResource(@CheckForNull App app, @Nonnull String resource)
       throws IOException {
-    if (app == null || !app.getAppStorageSource().equals(AppStorageSource.JCLOUDS)) {
+    if (app == null || !app.getAppStorageSource().equals(AppStorageSource.BLOB_STORE)) {
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java]: If the constant changes this `equals` might break for existing app entries, wouldn't it? 
So maybe we should have a method on the enum class that treats the new and the old name as both being equal?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java [hunk 4]
```diff
@@ -27,43 +27,38 @@
...
-  @Test
-  void verifyTransient() {
-    assertDoesNotThrow(() -> newBuilder("transient"));
+  @Bean
+  public BlobStoreService blobStoreService(
+      DhisConfigurationProvider configurationProvider, LocationManager locationManager) {
+    FileStoreProvider provider =
+        FileStoreProvider.of(
+            configurationProvider.getProperty(ConfigurationKey.FILESTORE_PROVIDER));
+    return switch (provider) {
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java:58]: If we have a configuration that decides what class we use for `BlobStoreService` then I would think that nobody should rely/inject that bean based on the bean name so giving each one an explicit name should not be necessary.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  @Override
+  @CheckForNull
+  public InputStream openStream(BlobKey key) {
+    Path p = resolve(key);
+    if (!Files.isRegularFile(p)) return null;
+    try {
+      return Files.newInputStream(p);
+    } catch (IOException e) {
+      log.warn("Unable to open stream for key: {}. {}", key, e.getMessage());
+      return null;
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:106]: I am not sure about the API contract here but returning `null` from a method that returns an `InputStream` seems odd and would likely just cause a NPE on the caller site. In that case I would think wrapping the `IOException` in `UncheckedIOException` is the better choice as it will maintain more information about the issue. Looking at the other methods they mostly seem to throw `UncheckedIOException`.

Taking one step further back I guess the interface contract should mention which exception to throw and what the conditions are so that all implementations can behave consistent so that callers can handle this independent of what implementation is used.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  @Override
+  public long contentLength(BlobKey key) {
+    Path p = resolve(key);
+    if (!Files.isRegularFile(p)) return 0L;
+    try {
+      return Files.size(p);
+    } catch (IOException e) {
+      return 0L;
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:117]: Similarly to above comment, I'd think returning zero is not necessarily an indication of an error so maybe this should be -1 or something?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    }
+  }
+
+  @Override
+  public void deleteDirectory(BlobKeyPrefix prefix) {
+    Path dir = baseDir.resolve(prefix.value());
+    if (!Files.isDirectory(dir)) return;
+    try (var paths = Files.walk(dir)) {
+      paths
+          .sorted(Comparator.reverseOrder())
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:153]: This seems odd. Why do we need it? What problem does it solve?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+      Files.createDirectories(dir);
+    } catch (IOException e) {
+      throw new UncheckedIOException("Unable to create directory " + dir, e);
+    }
+  }
+
+  @Override
+  @CheckForNull
+  public URI signedGetUri(BlobKey key, long expirationSeconds) {
+    return null;
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:225]: Feels like `null` has the role of `UnsupportedOperationExcerption`

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+  private void bulkDelete(String bucket, List<ObjectIdentifier> ids) {
+    DeleteObjectsResponse resp =
+        s3.deleteObjects(
+            DeleteObjectsRequest.builder()
+                .bucket(bucket)
+                .delete(d -> d.objects(ids).build())
+                .build());
+    if (!resp.errors().isEmpty()) {
+      S3Error first = resp.errors().get(0);
+      log.warn(
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java:440]: Maybe failed deletion should somehow tickle back to the caller of our blob service API so they can recognize when this happens.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+      InputStream content,
+      long contentLength,
+      @CheckForNull String contentType,
+      @CheckForNull ContentDisposition contentDisposition,
+      @CheckForNull ContentHash contentHash) {
+    // Read exactly contentLength bytes per BlobStoreService.putBlob's contract (the SDK-backed
+    // impls rely on the same guarantee for Content-Length). Math.toIntExact rejects anything
+    // larger than Integer.MAX_VALUE up-front — Transient is in-memory, so capping at int is the
+    // right limit anyway.
+    int length = Math.toIntExact(contentLength);
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java:109]: This means we already know that we cannot assume to store files larger than int range of bytes so maybe we should reflect that in the API already so the caller can decide how to deal with that?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+  @Override
+  public void deleteDirectory(BlobKeyPrefix prefix) {
+    String pfx = prefix.value() + "/";
+    blobs.keySet().removeIf(k -> k.startsWith(pfx));
+  }
+
+  @Override
+  public Iterable<BlobKeyPrefix> listFolders(BlobKeyPrefix prefix) {
+    String pfx = prefix.value() + "/";
+    Set<String> immediate = new HashSet<>();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java]: Maybe a `TreeSet` is a better choice, instead of random order bashed on hashes you get alphabetic order which is at least helpful during debug

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/storage/TransientBlobStoreServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import org.junit.jupiter.api.Test;
+
+/**
+ * Unit tests for {@link TransientBlobStoreService} edge cases that the contract test doesn't probe.
+ */
+class TransientBlobStoreServiceTest {
+
+  @Test
+  void putBlob_streamShorterThanContentLength_throws() {
+    TransientBlobStoreService svc = newService();
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/storage/TransientBlobStoreServiceTest.java]: Even if the test is specifically for one implementation I think it makes sense to type as if we go through the interface so we don't accidentally use class specific methods

**Review by jbee — COMMENTED** [review]
Some details to sort out:

* enum name change (migration needed?)
* details store API error contract (null vs exception, which exception, consistency)

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java (hunk 1 above)
Comment (david-mackessy) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java]: Yes, we need to know how this is used. I can't see it used anywhere in the code (other than setter/getter on App). I'll put out a question to some channels asking if any frontend app use this and if so for what purpose. I don't know why this would be present in the API.
It is only used in `App` object which is not persisted. No migration necessary for this.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java (hunk 2 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java]: sure

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java (hunk 3 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java]: I think I might just try to delete this enum, depending on the feedback received about its usage. It doesn't provide any value anymore. The history of having 2 values here `LOCAL` and `JCLOUDS` during a period of transition made sense, but not anymore.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java (hunk 5 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:106]: the interface method has info about returning null & `@CheckForNull` also used. It's pretty clear what to expect as a caller. But I can update it to throw `UncheckedIOException` instead of returning null for the 2 paths that currently return null if you'd like?

> returning null from a method that returns an InputStream seems odd and would likely just cause a NPE on the caller site

This is basically how Java works :) returning null from any object return type isn't odd behaviour in Java.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java (hunk 8 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:225]: pretty much, just a choice of `null` or throw an exception. Javadoc documents this.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java (hunk 9 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java:440]: That's an option. I think this matches existing behaviour.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java (hunk 10 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/TransientBlobStoreService.java:109]: i'll modify the javadoc for this with that info. It's only for tests really.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java (hunk 4 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java:58]: Not sure I understand this comment, maybe a chat?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java (hunk 3 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/BlobStoreAppStorageService.java]: `AppStorageSource` removed as no longer needed.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java (hunk 1 above)
Comment (david-mackessy) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppStorageSource.java]: `AppStorageSource` removed as no longer needed.

**Review by jbee — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java (hunk 4 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/BlobStoreConfig.java:58]: discussed and explained changes

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java (hunk 5 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/FileSystemBlobStoreService.java:106]: discussed and explained changes.

**Review by Philip-Larsen-Donnelly — APPROVED** [review]
The slow speed of the S3 uploads* makes this painful to test, but apart from that I didn't find any obvious issues with the storage.
Tested apps, icons, avatars and data values.
*Note - everything is slow. Even the icon fix patch too a couple of minutes to populate icons and the API call timed out after one minute.

**Comment by david-mackessy** [comment]
> The slow speed of the S3 uploads* makes this painful to test, but apart from that I didn't find any obvious issues with the storage. Tested apps, icons, avatars and data values. *Note - everything is slow. Even the icon fix patch too a couple of minutes to populate icons and the API call timed out after one minute.

This testing was extremely valuable @Philip-Larsen-Donnelly, thank you. How features use the BlobStore is very ad-hoc. When issues are known (like bundled app install or icon patch), these can be looked at & improved 👍

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+   *
+   * <p>The SDK's SigV4 chunked-payload signing pipeline runs inside {@code RetryableStage}: when a
+   * request is retried (transient 5xx / socket reset / etc.), the whole pipeline re-runs and {@code
+   * AwsChunkedV4PayloadSigner.sign} calls {@code ContentStreamProvider.newStream()} again on each
+   * attempt. The wrapper from {@link RequestBody#fromInputStream} can only honour the second call
+   * when the underlying stream supports {@code mark/reset} — for streams that don't (e.g. {@code
+   * ZipFile.getInputStream(zipEntry)} used by app install) the wrapper throws.
+   *
+   * <p>If the caller's stream is non-resetable we pre-buffer it into a byte array and use {@link
+   * RequestBody#fromBytes}, which is safely re-readable. Resetable streams (file streams, byte
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java:267]: After digging a bit (After some tips from Claude); I don't think file stream (FileInputStream) is actually resettable, as it doesn't implement the mark and reset methods. Not a critical thing per se, but it makes the next point more relevant:

If the stream is not resettable, we read the stream into a buffer, effectively reading the whole file into our heap. We might want to have a hard limit somewhere, either based on heap space or just a hard limit on individual file sizes; Cause now there's a way to upload files big enough to run us OOM, either with one big file, or with several, simultaneous medium files.

However, looking at "normal" use and expectations; And the fact that the user can set a sane file size limit in dhis.conf; I don't see this issue as a blocker, rather something to keep in mind; Maybe add to documentation?

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java (hunk 13 above)
Comment (david-mackessy) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/storage/S3BlobStoreService.java:267]: We can incorporate the system max file size here as a safeguard?
`max.file_upload_size` - default is 10MB.
At least then we don't interfere with system restrictions, which is probably what we want.
